File: apps/downloadx/src/meta.ts
import { META_EXT, META_SCHEMA_VERSION } from './constants.js';
import type {
  ChunkSnapshot,
  DownloadState,
  InjectedFunctions,
  MetaFile,
  ProbeResult,
} from './types.js';

/**
 * Sidecar JSON persisted under the cache directory:
 *   `{cachePath}/{id}.downloadx.json`
 *
 * The file is keyed by download id (not filename) so it can be written before
 * the probe completes — that lets `createDownloadX` rebuild its in-memory list
 * from disk on next startup, even for downloads that never started.
 *
 * Written atomically: write-to-tmp then rename, so a crash in the middle of
 * `persistMeta()` can't corrupt an existing valid meta file.
 */

const textEncoder = new TextEncoder();
const textDecoder = new TextDecoder('utf-8', { fatal: true });

export interface MetaLocator {
  /** Directory the meta JSON lives in (usually cachePath). */
  dir: string;
  /** Download id — the meta filename is `{id}.downloadx.json`. */
  id: string;
}

export function metaPath(io: InjectedFunctions, locator: MetaLocator): string {
  return io.joinPath(locator.dir, `${locator.id}${META_EXT}`);
}

function tmpPath(target: string): string {
  return `${target}.tmp`;
}

export interface CreateEmptyMetaInput {
  id: string;
  url: string;
  now?: () => number;
}

/**
 * Meta for a download that has been registered but not probed yet. Fields that
 * depend on the probe (filename, finalUrl, totalSize, etc.) start as null/zero
 * and are filled in by {@link applyProbeToMeta} once the probe completes.
 */
export function createEmptyMeta(input: CreateEmptyMetaInput): MetaFile {
  const ts = (input.now ?? Date.now)();
  return {
    schemaVersion: META_SCHEMA_VERSION,
    id: input.id,
    url: input.url,
    finalUrl: null,
    filename: null,
    totalSize: null,
    acceptsRanges: false,
    etag: null,
    lastModified: null,
    contentType: null,
    createdAt: ts,
    updatedAt: ts,
    state: 'idle',
    chunks: [],
    addedAt: ts,
    completedAt: null,
    errorMessage: null,
    speedLimit: null,
    targetChunkCount: null,
    targetPath: null,
    minChunkSize: null,
    journal: null,
  };
}

export interface CreateMetaInput {
  id: string;
  url: string;
  probe: ProbeResult;
  chunks: ChunkSnapshot[];
  now?: () => number;
}

/** Builds a fresh meta directly from a probe result (used when no prior meta exists). */
export function createMeta(input: CreateMetaInput): MetaFile {
  const meta = createEmptyMeta(
    input.now !== undefined
      ? { id: input.id, url: input.url, now: input.now }
      : { id: input.id, url: input.url },
  );
  return applyProbeToMeta(meta, input.probe, input.chunks);
}

/** Merges a probe result into an existing meta, returning the same object. */
export function applyProbeToMeta(
  meta: MetaFile,
  probe: ProbeResult,
  chunks: ChunkSnapshot[],
): MetaFile {
  meta.finalUrl = probe.finalUrl;
  meta.filename = probe.filename;
  meta.totalSize = probe.totalSize;
  meta.acceptsRanges = probe.acceptsRanges;
  meta.etag = probe.etag;
  meta.lastModified = probe.lastModified;
  meta.contentType = probe.contentType;
  meta.chunks = chunks;
  meta.updatedAt = Date.now();
  return meta;
}

export async function loadMeta(
  io: InjectedFunctions,
  locator: MetaLocator,
): Promise<MetaFile | null> {
  const path = metaPath(io, locator);
  if (!(await io.exists(path))) return null;
  try {
    const buf = await io.readFile(path);
    const text = textDecoder.decode(buf);
    const parsed: unknown = JSON.parse(text);
    const validated = validate(parsed);
    return validated;
  } catch {
    // Corrupt meta → treat as missing so a fresh download can start. The old
    // file is left on disk so the user can inspect it if needed.
    return null;
  }
}

/**
 * Scans `dir` for `*.downloadx.json` files and loads each one sequentially.
 * Corrupt or schema-mismatched files are skipped (left on disk). Sequential
 * instead of parallel because some injected I/O backends may not handle high
 * concurrency, and restore happens once at startup — latency isn't critical.
 */
export async function listMetaFiles(io: InjectedFunctions, dir: string): Promise<MetaFile[]> {
  if (!(await io.exists(dir))) return [];
  const entries = await io.listDir(dir);
  const out: MetaFile[] = [];
  for (const name of entries) {
    if (!name.endsWith(META_EXT)) continue;
    const id = name.slice(0, -META_EXT.length);
    const meta = await loadMeta(io, { dir, id });
    if (meta !== null) out.push(meta);
  }
  return out;
}

export async function persistMeta(
  io: InjectedFunctions,
  locator: MetaLocator,
  meta: MetaFile,
): Promise<void> {
  await io.mkdir(locator.dir);
  const target = metaPath(io, locator);
  const tmp = tmpPath(target);
  const payload: MetaFile = { ...meta, updatedAt: Date.now() };
  const encoded = textEncoder.encode(JSON.stringify(payload, null, 2));
  await io.writeFile(tmp, encoded);
  await io.rename(tmp, target);
}

export async function deleteMeta(io: InjectedFunctions, locator: MetaLocator): Promise<void> {
  const target = metaPath(io, locator);
  await io.unlink(target);
}

/** Updates an existing meta object in place and returns it for chaining. */
export function updateMeta(
  meta: MetaFile,
  patch: Partial<
    Pick<
      MetaFile,
      | 'state'
      | 'chunks'
      | 'speedLimit'
      | 'targetChunkCount'
      | 'targetPath'
      | 'completedAt'
      | 'errorMessage'
    >
  >,
): MetaFile {
  if (patch.state !== undefined) meta.state = patch.state;
  if (patch.chunks !== undefined) meta.chunks = patch.chunks;
  if ('speedLimit' in patch) meta.speedLimit = patch.speedLimit ?? null;
  if ('targetChunkCount' in patch) meta.targetChunkCount = patch.targetChunkCount ?? null;
  if ('targetPath' in patch) meta.targetPath = patch.targetPath ?? null;
  if ('completedAt' in patch) meta.completedAt = patch.completedAt ?? null;
  if ('errorMessage' in patch) meta.errorMessage = patch.errorMessage ?? null;
  meta.updatedAt = Date.now();
  return meta;
}

/**
 * Decides whether a meta file represents the same remote resource a fresh
 * probe describes. Compares ETag first (strong), then Last-Modified, then
 * total size. If none match confidently, we can't trust resume → return false.
 */
export function canResumeAgainst(meta: MetaFile, probe: ProbeResult): boolean {
  if (meta.schemaVersion !== META_SCHEMA_VERSION) return false;
  if (meta.totalSize !== probe.totalSize) return false;
  if (meta.etag !== null && probe.etag !== null) {
    return meta.etag === probe.etag;
  }
  if (meta.lastModified !== null && probe.lastModified !== null) {
    return meta.lastModified === probe.lastModified;
  }
  // No validators — size match is the only signal we have. It's weak, but
  // better than forcing a restart every time.
  return true;
}

/** Maps a raw download state onto what should be persisted on pause/crash. */
export function dehydrateState(state: DownloadState): DownloadState {
  // `downloading` is never durable — if we crashed we were paused.
  if (state === 'downloading' || state === 'probing') return 'paused';
  return state;
}

function validate(value: unknown): MetaFile {
  if (typeof value !== 'object' || value === null) throw new Error('meta: not an object');
  const v = value as Record<string, unknown>;
  if (v['schemaVersion'] !== META_SCHEMA_VERSION) {
    throw new Error(`meta: unsupported schemaVersion ${String(v['schemaVersion'])}`);
  }
  assertString(v, 'id');
  assertString(v, 'url');
  assertNullableString(v, 'finalUrl');
  assertNullableString(v, 'filename');
  assertNullableNumber(v, 'totalSize');
  assertBoolean(v, 'acceptsRanges');
  assertNullableString(v, 'etag');
  assertNullableString(v, 'lastModified');
  assertNullableString(v, 'contentType');
  assertNumber(v, 'createdAt');
  assertNumber(v, 'updatedAt');
  assertString(v, 'state');
  if (!Array.isArray(v['chunks'])) throw new Error('meta: chunks must be array');
  const chunks = v['chunks'].map((c, i) => validateChunk(c, i));
  assertNumber(v, 'addedAt');
  assertNullableNumber(v, 'completedAt');
  assertNullableString(v, 'errorMessage');
  assertNullableNumber(v, 'speedLimit');
  assertNullableNumber(v, 'targetChunkCount');
  assertNullableString(v, 'targetPath');
  const minChunkSize = 'minChunkSize' in v ? v['minChunkSize'] : null;
  const journal = 'journal' in v ? v['journal'] : null;
  return {
    ...(v as unknown as MetaFile),
    chunks,
    minChunkSize: typeof minChunkSize === 'number' ? minChunkSize : null,
    journal: typeof journal === 'boolean' ? journal : null,
  };
}

function validateChunk(value: unknown, index: number): ChunkSnapshot {
  if (typeof value !== 'object' || value === null) {
    throw new Error(`meta: chunk[${index}] not an object`);
  }
  const c = value as Record<string, unknown>;
  assertString(c, 'id');
  assertNumber(c, 'offset');
  assertNumber(c, 'length');
  assertNumber(c, 'downloadedBytes');
  assertString(c, 'status');
  assertString(c, 'quality');
  assertNumber(c, 'retries');
  return value as ChunkSnapshot;
}

function assertString(obj: Record<string, unknown>, key: string): void {
  if (typeof obj[key] !== 'string') throw new Error(`meta: ${key} must be string`);
}
function assertNullableString(obj: Record<string, unknown>, key: string): void {
  const v = obj[key];
  if (v !== null && typeof v !== 'string') throw new Error(`meta: ${key} must be string|null`);
}
function assertNumber(obj: Record<string, unknown>, key: string): void {
  if (typeof obj[key] !== 'number') throw new Error(`meta: ${key} must be number`);
}
function assertNullableNumber(obj: Record<string, unknown>, key: string): void {
  const v = obj[key];
  if (v !== null && typeof v !== 'number') throw new Error(`meta: ${key} must be number|null`);
}
function assertBoolean(obj: Record<string, unknown>, key: string): void {
  if (typeof obj[key] !== 'boolean') throw new Error(`meta: ${key} must be boolean`);
}

File: apps/downloadx_dart/lib/src/meta.dart
import 'dart:convert';

import 'constants.dart';
import 'io.dart';
import 'types.dart';

/// Sidecar JSON persisted under the cache directory:
///   `{cachePath}/{id}.downloadx.json`
///
/// Keyed by download id (not filename) so it can be written before the probe
/// completes. Written atomically: write-to-tmp then rename, so a crash mid
/// [persistMeta] can't corrupt an existing valid meta file.

const _jsonEncoder = JsonEncoder.withIndent('  ');

/// Locates the sidecar meta file for a download on disk.
class MetaLocator {
  /// Directory the meta JSON lives in (usually cachePath).
  final String dir;

  /// Download id — the meta filename is `{id}.downloadx.json`.
  final String id;

  /// Creates a [MetaLocator].
  const MetaLocator({required this.dir, required this.id});
}

/// Returns the absolute path to the meta file described by [locator].
String metaPath(DownloadxIo io, MetaLocator locator) =>
    io.joinPath([locator.dir, '${locator.id}$metaExt']);

String _tmpPath(String target) => '$target.tmp';

int _defaultNow() => DateTime.now().millisecondsSinceEpoch;

/// Meta for a download that has been registered but not probed yet. Probe
/// fields start null/zero and are filled by [applyProbeToMeta].
MetaFile createEmptyMeta(
    {required String id, required String url, int Function()? now}) {
  final ts = (now ?? _defaultNow)();
  return MetaFile(
    schemaVersion: metaSchemaVersion,
    id: id,
    url: url,
    finalUrl: null,
    filename: null,
    totalSize: null,
    acceptsRanges: false,
    etag: null,
    lastModified: null,
    contentType: null,
    createdAt: ts,
    updatedAt: ts,
    state: DownloadState.idle,
    chunks: [],
    addedAt: ts,
    completedAt: null,
    errorMessage: null,
    speedLimit: null,
    targetChunkCount: null,
    targetPath: null,
    minChunkSize: null,
    journal: null,
  );
}

/// Builds a fresh meta directly from a probe result.
MetaFile createMeta({
  required String id,
  required String url,
  required ProbeResult probe,
  required List<ChunkSnapshot> chunks,
  int Function()? now,
}) {
  final meta = createEmptyMeta(id: id, url: url, now: now);
  return applyProbeToMeta(meta, probe, chunks);
}

/// Merges a probe result into an existing meta, returning the same object.
MetaFile applyProbeToMeta(
    MetaFile meta, ProbeResult probe, List<ChunkSnapshot> chunks) {
  meta.finalUrl = probe.finalUrl;
  meta.filename = probe.filename;
  meta.totalSize = probe.totalSize;
  meta.acceptsRanges = probe.acceptsRanges;
  meta.etag = probe.etag;
  meta.lastModified = probe.lastModified;
  meta.contentType = probe.contentType;
  meta.isHls = probe.isHls;
  meta.chunks = chunks;
  meta.updatedAt = _defaultNow();
  return meta;
}

/// Loads and validates a meta file. Returns null when the file is missing or corrupt.
Future<MetaFile?> loadMeta(DownloadxIo io, MetaLocator locator) async {
  final path = metaPath(io, locator);
  if (!await io.exists(path)) return null;
  try {
    final buf = await io.readFile(path);
    final text = utf8.decode(buf);
    final parsed = jsonDecode(text);
    return _validate(parsed);
  } catch (_) {
    // Corrupt meta → treat as missing so a fresh download can start. The old
    // file is left on disk for inspection.
    return null;
  }
}

/// Scans [dir] for `*.downloadx.json` and loads each sequentially. Corrupt or
/// schema-mismatched files are skipped (left on disk).
Future<List<MetaFile>> listMetaFiles(DownloadxIo io, String dir) async {
  if (!await io.exists(dir)) return [];
  final entries = await io.listDir(dir);
  final out = <MetaFile>[];
  for (final name in entries) {
    if (!name.endsWith(metaExt)) continue;
    final id = name.substring(0, name.length - metaExt.length);
    final meta = await loadMeta(io, MetaLocator(dir: dir, id: id));
    if (meta != null) out.add(meta);
  }
  return out;
}

/// Atomically writes [meta] to disk (write-to-tmp then rename).
Future<void> persistMeta(
    DownloadxIo io, MetaLocator locator, MetaFile meta) async {
  await io.mkdir(locator.dir);
  final target = metaPath(io, locator);
  final tmp = _tmpPath(target);
  meta.updatedAt = _defaultNow();
  final encoded = utf8.encode(_jsonEncoder.convert(meta.toJson()));
  await io.writeFile(tmp, encoded);
  await io.rename(tmp, target);
}

/// Deletes the meta file. Silently no-ops when the file does not exist.
Future<void> deleteMeta(DownloadxIo io, MetaLocator locator) async {
  await io.unlink(metaPath(io, locator));
}

/// Decides whether a meta file represents the same remote resource a fresh
/// probe describes. ETag first (strong), then Last-Modified, then total size.
bool canResumeAgainst(MetaFile meta, ProbeResult probe) {
  if (meta.schemaVersion != metaSchemaVersion) return false;
  if (meta.totalSize != probe.totalSize) return false;
  if (meta.etag != null && probe.etag != null) {
    return meta.etag == probe.etag;
  }
  if (meta.lastModified != null && probe.lastModified != null) {
    return meta.lastModified == probe.lastModified;
  }
  // No validators — size match is the only (weak) signal we have.
  return true;
}

/// Maps a raw download state onto what should be persisted on pause/crash.
DownloadState dehydrateState(DownloadState state) {
  // `downloading`/`probing` are never durable — if we crashed we were paused.
  if (state == DownloadState.downloading || state == DownloadState.probing) {
    return DownloadState.paused;
  }
  return state;
}

// ---------------------------------------------------------------------------
// Validation
// ---------------------------------------------------------------------------

MetaFile _validate(dynamic value) {
  if (value is! Map) throw const FormatException('meta: not an object');
  final v = value.cast<String, dynamic>();
  if (v['schemaVersion'] != metaSchemaVersion) {
    throw FormatException(
        'meta: unsupported schemaVersion ${v['schemaVersion']}');
  }
  _assertString(v, 'id');
  _assertString(v, 'url');
  _assertNullableString(v, 'finalUrl');
  _assertNullableString(v, 'filename');
  _assertNullableNumber(v, 'totalSize');
  _assertBoolean(v, 'acceptsRanges');
  _assertNullableString(v, 'etag');
  _assertNullableString(v, 'lastModified');
  _assertNullableString(v, 'contentType');
  _assertNumber(v, 'createdAt');
  _assertNumber(v, 'updatedAt');
  _assertString(v, 'state');
  if (v['chunks'] is! List) {
    throw const FormatException('meta: chunks must be array');
  }
  final chunks = <ChunkSnapshot>[];
  final rawChunks = v['chunks'] as List;
  for (var i = 0; i < rawChunks.length; i += 1) {
    chunks.add(_validateChunk(rawChunks[i], i));
  }
  _assertNumber(v, 'addedAt');
  _assertNullableNumber(v, 'completedAt');
  _assertNullableString(v, 'errorMessage');
  _assertNullableNumber(v, 'speedLimit');
  _assertNullableNumber(v, 'targetChunkCount');
  _assertNullableString(v, 'targetPath');
  final minChunkSize = v['minChunkSize'];
  final journal = v['journal'];
  final isHls = v['isHls'];
  final description = v['description'];
  final metadata = v['metadata'];

  return MetaFile(
    schemaVersion: metaSchemaVersion,
    id: v['id'] as String,
    url: v['url'] as String,
    finalUrl: v['finalUrl'] as String?,
    filename: v['filename'] as String?,
    totalSize: (v['totalSize'] as num?)?.toInt(),
    acceptsRanges: v['acceptsRanges'] as bool,
    etag: v['etag'] as String?,
    lastModified: v['lastModified'] as String?,
    contentType: v['contentType'] as String?,
    createdAt: (v['createdAt'] as num).toInt(),
    updatedAt: (v['updatedAt'] as num).toInt(),
    state: downloadStateFromString(v['state'] as String),
    chunks: chunks,
    addedAt: (v['addedAt'] as num).toInt(),
    completedAt: (v['completedAt'] as num?)?.toInt(),
    errorMessage: v['errorMessage'] as String?,
    speedLimit: (v['speedLimit'] as num?)?.toInt(),
    targetChunkCount: (v['targetChunkCount'] as num?)?.toInt(),
    targetPath: v['targetPath'] as String?,
    minChunkSize: minChunkSize is num ? minChunkSize.toInt() : null,
    journal: journal is bool ? journal : null,
    isHls: isHls is bool ? isHls : false,
    description: description is String ? description : null,
    metadata: _asStringMap(metadata),
  );
}

/// Returns a `Map<String, String>` when [value] is a plain map whose values are
/// all strings, otherwise null. Used for the optional `metadata` field.
Map<String, String>? _asStringMap(dynamic value) {
  if (value is! Map) return null;
  final out = <String, String>{};
  for (final entry in value.entries) {
    if (entry.key is! String || entry.value is! String) return null;
    out[entry.key as String] = entry.value as String;
  }
  return out;
}

ChunkSnapshot _validateChunk(dynamic value, int index) {
  if (value is! Map) throw FormatException('meta: chunk[$index] not an object');
  final c = value.cast<String, dynamic>();
  _assertString(c, 'id');
  _assertNumber(c, 'offset');
  _assertNumber(c, 'length');
  _assertNumber(c, 'downloadedBytes');
  _assertString(c, 'status');
  _assertString(c, 'quality');
  _assertNumber(c, 'retries');
  // Optional HLS segment fields — validated only when present so old metas and
  // normal chunks remain valid.
  if (c.containsKey('isSegment') && c['isSegment'] != null && c['isSegment'] is! bool) {
    throw FormatException('meta: chunk[$index].isSegment must be boolean');
  }
  if (c.containsKey('targetFilePath') &&
      c['targetFilePath'] != null &&
      c['targetFilePath'] is! String) {
    throw FormatException('meta: chunk[$index].targetFilePath must be string');
  }
  if (c.containsKey('uri') && c['uri'] != null && c['uri'] is! String) {
    throw FormatException('meta: chunk[$index].uri must be string');
  }
  if (c.containsKey('durationSec') && c['durationSec'] != null && c['durationSec'] is! num) {
    throw FormatException('meta: chunk[$index].durationSec must be number');
  }
  return ChunkSnapshot.fromJson(c);
}

void _assertString(Map<String, dynamic> o, String key) {
  if (o[key] is! String) throw FormatException('meta: $key must be string');
}

void _assertNullableString(Map<String, dynamic> o, String key) {
  final v = o[key];
  if (v != null && v is! String) {
    throw FormatException('meta: $key must be string|null');
  }
}

void _assertNumber(Map<String, dynamic> o, String key) {
  if (o[key] is! num) throw FormatException('meta: $key must be number');
}

void _assertNullableNumber(Map<String, dynamic> o, String key) {
  final v = o[key];
  if (v != null && v is! num) {
    throw FormatException('meta: $key must be number|null');
  }
}

void _assertBoolean(Map<String, dynamic> o, String key) {
  if (o[key] is! bool) throw FormatException('meta: $key must be boolean');
}
